Add details about detecting CPU architecture (#2437)
MicrosoftDocs/edge-developer: microsoft-edge/web-platform/how-to-detect-win11.md

---

> title: Detect Windows 11 and CPU Architecture using User-Agent Client Hints

Reviewer: ```suggestion
title: Detect Windows 11 and CPU architecture using User-Agent Client Hints
```

description: How to differentiate Windows 10 and Windows 11 and CPU architecture using User-Agent Client Hints.
author: MSEdgeTeam
ms.author: msedgedevrel
ms.topic: conceptual
ms.prod: microsoft-edge
ms.date: 11/22/2022
---

> # Detect Windows 11 and CPU Architecture using User-Agent Client Hints

Reviewer: ```suggestion
# Detect Windows 11 and CPU architecture using User-Agent Client Hints
```

For details about these two approaches, see [Detecting Microsoft Edge from your website](user-agent-guidance.md).

> In Microsoft Edge (and also in Google Chrome), sites can differentiate between users on Windows 11 and Windows 10, and the CPU architecture of the device, via User-Agent Client Hints (UA-CH).  This information can be found in the following UA-CH request headers:

Reviewer: ```suggestion
In Microsoft Edge (and also in Google Chrome), sites can differentiate between users on Windows 11 and Windows 10, and can detect the CPU architecture of the device, via User-Agent Client Hints (UA-CH).  This information can be found in the following UA-CH request headers:
```

| Header field | Values that indicate Windows 10 | Values that indicate Windows 11 |
| --- | --- | --- |
| `Sec-CH-UA-Platform` | `Windows` | `Windows` |
| `Sec-CH-UA-Platform-Version` | values between `1.0.0` and `10.0.0` | `13.0.0` and above |

> User-Agent strings won't be updated to differentiate between Windows 11 and Windows 10 or between CPU architectures.  We don't recommend using User-Agent strings to retrieve user agent data.  Browsers that don't support User-Agent Client Hints won't be able to differentiate between Windows 11 and Windows 10, or between CPU architectures.

Reviewer: ```suggestion
User-Agent strings won't be updated to differentiate between Windows 11 and Windows 10, or to differentiate between CPU architectures.  We don't recommend using User-Agent strings to retrieve user agent data.  Browsers that don't support User-Agent Client Hints won't be able to differentiate between Windows 11 and Windows 10, or between CPU architectures.
```

```

<!-- ====================================================================== -->
## Sample code for detecting ARM or x86

> Detection of CPU architecture is helpful for websites that wish to guide visitors to download applications built for the CPU in their system. This is particularly helpful for ARM-based devices so that customers download the native ARM version of an application, instead of inadvertently installing an app built for x86 and experiencing reduced performance from emulation.

Reviewer: ```suggestion
Use detection of CPU architecture to have your website automatically download the version of your app that's built specifically for the user's CPU.

CPU detection is particularly helpful for ARM-based devices, so that a customer using an ARM device automatically downloads the native ARM version of an application.  This prevents the user from inadvertently installing an app that's built for x86, and then experiencing reduced performance due to emulation.
```
Writer pass:
* broke up long sentence per Acrolinx.
* singular instead of plural, to reduce ambiguity.

The following code detects CPU architecture:

```javascript
navigator.userAgentData.getHighEntropyValues(["architecture","bitness"])
 .then(ua => {
   if (navigator.userAgentData.platform === "Windows") {
     if (ua.architecture === 'x86') {
       if (ua.bitness === '64') {
         console.log("x86_64");
       }
       else if (ua.bitness === '32') {
         console.log("x86");
       }
     }
     else if (ua.architecture === 'arm') {
       if (ua.bitness === '64') {
         console.log("ARM64");
       }
       else if (ua.bitness === '32') {
         console.log("ARM32");
       }
     }
   }
   else {
     console.log("Not running on Windows");
   }
 });

> 

Reviewer: ```suggestion
```

```

<!-- ====================================================================== -->
## Sample code for detecting ARM or x86

Use detection of CPU architecture to have your website automatically download the version of your app that's built specifically for the user's CPU.

> CPU detection is particularly helpful for ARM-based devices, so that a customer using an ARM device automatically downloads the native ARM version of an application.  This prevents the user from inadvertently installing an app that's built for x86, and then experiencing reduced performance due to emulation.

Reviewer: ```suggestion
CPU detection is particularly helpful for ARM-based devices, so that a customer using an ARM device automatically downloads the native ARM version of the application.  This prevents the user from inadvertently installing an app that's built for x86, and then experiencing reduced performance due to emulation.
```
